### PR TITLE
Consume all pages when counting

### LINKED ISSUE #243: count(...) is broken when there are too many items.
tl dr: The count function is broken when there are many items per hash key.  It is likely that it is returning the count for only the first page.  

I'm getting a different number from pynamo count than from using boto or from querying for all items and counting them locally:
```
>>> NumberPoolServiceIndex.count(hash_key='number_pool')
8528
>>> dynamodb_table.query_count(index=settings.DYNAMODB_NUMBER_POOL_TABLE_SERVICE_INDEX, s__eq='number_pool')
10175
>>> len(list(NumberPoolServiceIndex.query('number_pool')))
10175
```

Here is the log when getting the count:
```
>>> NumberPoolServiceIndex.count(hash_key='number_pool')
DEBUG:pynamodb.connection.base:Calling Query with arguments {'IndexName': 'switchboard-production-number-pool-service-index', 'ReturnConsumedCapacity': 'TOTAL', 'TableName': 'switchboard-production-number-pool', 'KeyConditions': {u's': {'ComparisonOperator': 'EQ', 'AttributeValueList': [{u'S': u'number_pool'}]}}, 'Select': 'COUNT'}
DEBUG:pynamodb.connection.base: Query consumed 73.0 units
8528
```

and here is the log for querying for all items:
```
>>> len(list(NumberPoolServiceIndex.query('number_pool')))
DEBUG:pynamodb.models:Fetching first query page
DEBUG:pynamodb.connection.base:Calling Query with arguments {'IndexName': 'switchboard-production-number-pool-service-index', 'TableName': 'switchboard-production-number-pool', 'ReturnConsumedCapacity': 'TOTAL', 'KeyConditions': {u's': {'ComparisonOperator': 'EQ', 'AttributeValueList': [{u'S': u'number_pool'}]}}}
DEBUG:pynamodb.connection.base: Query consumed 73.0 units
DEBUG:pynamodb.models:Fetching query page with exclusive start key: {u's': {u'S': u'number_pool'}, u'lu': {u'N': u'1489565278284'}, u'pn': {u'S': u'+15015749931'}}
DEBUG:pynamodb.connection.base:Calling Query with arguments {'ExclusiveStartKey': {u's': {u'S': u'number_pool'}, u'lu': {u'N': u'1489565278284'}, u'pn': {u'S': u'+15015749931'}}, 'IndexName': 'switchboard-production-number-pool-service-index', 'TableName': 'switchboard-production-number-pool', 'ReturnConsumedCapacity': 'TOTAL', 'KeyConditions': {u's': {'ComparisonOperator': 'EQ', 'AttributeValueList': [{u'S': u'number_pool'}]}}}
DEBUG:pynamodb.connection.base: Query consumed 14.5 units
10182
```

It appears that there are multiple pages and its quite likely that the count function is just returning the count for the first page.

### PR DESCRIPTION
**Pull request by garrettheel**
Fixes #243

### CODE REVIEW

**Comment by jmphilli** [comment]
tests?

**Comment by lita** [comment]
can you add a test? 😹

File: pynamodb/tests/test_model.py [hunk 1]
```diff
@@ -1477,6 +1477,40 @@ def test_index_count(self):
             }
             deep_eq(args, params, _assert=True)
 
+    def test_index_multipage_count(self):
```
Comment (lita) [inline pynamodb/tests/test_model.py]: Can we also add a test for when the data is None?

File: pynamodb/tests/test_model.py [hunk 2]
```diff
@@ -1477,6 +1477,40 @@ def test_index_count(self):
...
+            last_evaluated_key = {
+                'user_name': {'S': u'user'},
+                'user_id': {'S': '1234'},
+            }
+            req.side_effect = [
+                {'Count': 1000, 'LastEvaluatedKey': last_evaluated_key},
+                {'Count': 42}
+            ]
+            res = CustomAttrNameModel.uid_index.count('foo')
+            self.assertEqual(res, 1042)
```
Comment (lita) [inline pynamodb/tests/test_model.py]: nit: 
```
{
    'Count': 42
}
```

**Comment by lita** [comment]
looks good other than adding that extra test. +1

File: pynamodb/tests/test_model.py (hunk 1 above)
Comment (garrettheel) [inline pynamodb/tests/test_model.py]: Do you mean when `LastEvaluatedKey` is None on the first page? If so, this is covered by the previous test

File: pynamodb/models.py [hunk 3]
```diff
@@ -542,16 +542,24 @@ def count(cls,
...
-            key_conditions=key_conditions,
-            query_filters=query_filters,
-            limit=limit,
-            select=COUNT
-        )
-        return data.get(CAMEL_COUNT)
+        count_buffer = 0
+        last_evaluated_key = None
+        data = None
+        while not data or last_evaluated_key:
```
Comment (mosoto) [inline pynamodb/models.py]: probably a stupid question but because I'm not aware of the output of query I'll ask it anyway - will query always return an object?  I'm wondering about the case where there are no items and I want to make sure this won't loop forever because of that "not data".

Comment (garrettheel) [inline pynamodb/models.py]: The most minimal object that should be returned is `{'Count': 0}`, which should be truth-y here. This is a good point though, so I replaced this check with an explicit `started` bool to be more explicit.

**Comment by mosoto** [comment]
+1

File: pynamodb/models.py [hunk 4]
```diff
@@ -542,16 +542,25 @@ def count(cls,
...
-            key_conditions=key_conditions,
-            query_filters=query_filters,
-            limit=limit,
-            select=COUNT
-        )
-        return data.get(CAMEL_COUNT)
+        count_buffer = 0
+        last_evaluated_key = None
+        started = False
+        while not started or last_evaluated_key:
```
Comment (jmphilli) [inline pynamodb/models.py]: can't you save the `started` var if you just set `last_evaluated_key` to a sentinel value?

Comment (garrettheel) [inline pynamodb/models.py]: Yeah, but I see this as a little hacky if there are two "null" states

File: pynamodb/tests/test_model.py (hunk 1 above)
Comment (lita) [inline pynamodb/tests/test_model.py]: Ah, yup. you are right.
